### ISSUE TITLE
box blows recursion stack

### ISSUE THREAD

**Issue by wimglenn** [cdgriffith/Box#20]
"box it up" doesn't handle recursion gracefully 

```
>>> from box import Box
>>> d = {}
>>> d['d'] = d
>>> b = Box(d)
>>> b.box_it_up()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/wglenn/.virtualenvs/scratch/lib/python3.6/site-packages/box.py", line 251, in box_it_up
    self[k].box_it_up()
  File "/home/wglenn/.virtualenvs/scratch/lib/python3.6/site-packages/box.py", line 251, in box_it_up
    self[k].box_it_up()
  File "/home/wglenn/.virtualenvs/scratch/lib/python3.6/site-packages/box.py", line 251, in box_it_up
    self[k].box_it_up()
  [Previous line repeated 978 more times]
  File "/home/wglenn/.virtualenvs/scratch/lib/python3.6/site-packages/box.py", line 250, in box_it_up
    if hasattr(self[k], 'box_it_up'):
  File "/home/wglenn/.virtualenvs/scratch/lib/python3.6/site-packages/box.py", line 317, in __getitem__
    return self.__convert_and_store(item, value)
  File "/home/wglenn/.virtualenvs/scratch/lib/python3.6/site-packages/box.py", line 332, in __convert_and_store
    **self.__box_config())
  File "/home/wglenn/.virtualenvs/scratch/lib/python3.6/site-packages/box.py", line 218, in __init__
    if isinstance(args[0], Mapping):
  File "/home/wglenn/.virtualenvs/scratch/lib64/python3.6/abc.py", line 182, in __instancecheck__
    if subclass in cls._abc_cache:
  File "/home/wglenn/.virtualenvs/scratch/lib64/python3.6/_weakrefset.py", line 75, in __contains__
    return wr in self.data
RecursionError: maximum recursion depth exceeded in comparison

```

Now `b.to_dict()` is broken.  

It also mangles the string representation afterwards : 

```
>>> print(b)
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d'
: {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': 
{'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'
d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {'d': {...}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}
}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}


```

**Comment by cdgriffith** [comment]
Thanks @wimglenn, great catch. 

Need to add config item that will define if this box has already been boxed up, so any sub box at any position won't run into this.

**Comment by cdgriffith** [comment]
I think I have a solid fix for the `to_dict`, and standard dictionary operations issues.  https://github.com/cdgriffith/Box/commit/7b28aa033eab6b35733049022e436910a269dfa7

```python
from box import Box
a = {}
a['a'] = a
bx = Box(a)
bx.a.a
Out[4]: <Box: {'a': {...}}>
bx.a.a.a
Out[5]: <Box: {'a': {...}}>
bx.a.a.a.to_dict()
Out[6]: {'a': {...}}
```

Which will raise the standard errors encounters with dicts that have circular references. 

```python
import json
json.dumps(a)
Traceback (most recent call last):
...
ValueError: Circular reference detected

bx.to_json()
Traceback (most recent call last):
...
ValueError: Circular reference detected


c = {}
c['a'] = c
c == a 
Traceback (most recent call last):
...
RecursionError: maximum recursion depth exceeded in comparison

bx == a 
Traceback (most recent call last):
...
RecursionError: maximum recursion depth exceeded in comparison
```

Though I am starting to think that `box_it_up` should simply raise an error like `to_json` or something would. Because it is saying "create all children into box objects", and just like the JSON circular reference error, you can't create an infinite amount of yourselves.  I would like to detect and at minimum throw that as a BoxError so it's assumed known behavior. Though I am not sure of a better way to do that then catch the RecurisonError (as the circular reference may be multiple levels between itself, can't just do a simple check for itself among it's values).

**Comment by cdgriffith** [comment]
After looking at how other similar libraries work for circular references, they seem to simply blow up as well.  So I will have to do more research into how best to handle this, from leave it as is to complex fix, and created the other Issue for the long term research and for making custom errors on detection. However the dev branch does at least can handle single level self references like this issue describes.

**Comment by cdgriffith** [comment]
First level circular references fixed in 3.0.1 https://github.com/cdgriffith/Box/tree/3.0.1
